File: Server/EMS-server/src/main/java/Packet.java
import com.google.gson.Gson;

import java.util.Map;

public class Packet {
    private final Map<String, Object> data;

    // Gson instance for JSON serialization/deserialization
    private static final Gson gson = new Gson();

    // Constructor
    public Packet(Map<String, Object> data) {
        this.data = data;
    }

    // Convert the packet to a JSON string
    public String data() {
        return gson.toJson(this.data);
    }

    @Override
    public String toString() {
        return this.data();
    }

    // Static method to create a Packet from a JSON string
    public static Packet fromString(String jsonString) {
        Map<String, Object> data = gson.fromJson(jsonString, Map.class);
        return new Packet(data);
    }

    // Get a value from the packet by key
    public Object get(String key) {
        Object val = this.data.get(key);

        if (val instanceof Double) return (int)(val.toString().charAt(0)) - 48;
        return this.data.get(key);
    }

    // Example usage
    public static void main(String[] args) {
        // Create a packet
        Packet packet = new Packet(Map.of("sensor", "temperature", "value", 22.5, "unit", "Celsius"));

        // Convert to JSON string
        String packetString = packet.data();
        System.out.println("Serialized Packet: " + packetString);

        // Reconstruct packet from JSON string
        Packet reconstructedPacket = Packet.fromString(packetString);
        System.out.println("Sensor: " + reconstructedPacket.get("sensor"));
        System.out.println("Value: " + reconstructedPacket.get("value"));
    }
}

File: Client/EMS-client/packet.py
import json

class Packet:
    def __init__(self, data):
        """
        Initialize a packet with a dictionary of key-value pairs.
        """
        if not isinstance(data, dict):
            raise TypeError("Packet data must be a dictionary.")
        self.data = data

    def __str__(self):
        """
        Convert the packet to a JSON string.
        """
        return json.dumps(self.data)

    @classmethod
    def parse(cls, string):
        """
        Create a Packet object from a JSON string.
        """
        data = json.loads(str(string))
        return cls(data)

    def __getitem__(self, key):
        """
        Allow direct access to packet data via key lookup.
        """
        return self.data[key]

    # Example usage
if __name__ == "__main__":
    # Create a packet
    packet = Packet({"sensor": "temperature", "value": 22.5, "unit": "Celsius"})

    # Convert to JSON string
    packet_string = packet
    print(f"Serialized Packet: {packet_string}")

    # Reconstruct packet from JSON string
    reconstructed_packet = Packet.parse(packet_string)
    print(f"Sensor: {reconstructed_packet['sensor']}")
    print(f"Value: {reconstructed_packet['value']}")
